src/ejabberd_config.erl

```diff
@@ -1219,109 +659,54 @@ get_local_config() ->
-                                             KeyPos, NewTerms, ValuePos),
-                                     {[], []}, OldTerms),
-    ToStart = lists:foldl(pa:bind(fun find_modules_to_start/4,
-                                  KeyPos, OldTerms), [], NewTerms),
-    #compare_result{to_start  = ToStart,
-                    to_stop   = ToStop,
-                    to_reload = ToReload}.
-
-find_modules_to_start(KeyPos, OldTerms, Element, ToStart) ->
-    case lists:keyfind(element(KeyPos, Element), KeyPos, OldTerms) of
-        false -> [Element | ToStart];
-        _ -> ToStart
-    end.
+%% @doc Returns current all options in memory, grouped by category.
+get_categorized_options() ->
+    Config = get_global_config(),
+    Local = get_local_config(),
+    HostsLocal = get_host_local_config(),
+    mongoose_config:make_categorized_options(Config, Local, HostsLocal).
 
-find_modules_to_change(KeyPos, NewTerms, ValuePos,
-                       Element, {ToStop, ToReload}) ->
-    case lists:keyfind(element(KeyPos, Element), KeyPos, NewTerms) of
-        false ->
-            {[Element | ToStop], ToReload};
-        NewElement ->
-            OldVal = element(ValuePos, Element),
-            NewVal = element(ValuePos, NewElement),
-            case OldVal == NewVal of
-                true ->
-                    {ToStop, ToReload};
-                false ->
-                    %% add also old value
-                    {ToStop,
-                     [{element(KeyPos, Element), OldVal, NewVal} | ToReload]}
-            end
+%% @doc Returns configs on disc and in memory for this node.
+%% This function prepares all state data to pass into pure code part
+%% (i.e. mongoose_config).
+config_state() ->
+    ConfigFile = get_ejabberd_config_path(),
+    Terms = get_plain_terms_file(ConfigFile),
+    #{mongoose_node => node(),
+      config_file => ConfigFile,
+      loaded_categorized_options => get_categorized_options(),
+      ondisc_config_terms => Terms}.
+
+%% @doc Returns config states from all nodes in cluster
+%% State from the local node comes as head of a list
+config_states() ->
+    ThisNodeState = config_state(),
+    Nodes = other_cluster_nodes(),
+    {S, F} = rpc:multicall(Nodes, ?MODULE, config_state, [], 30000),
+    case F of
+        [] ->
+            [ThisNodeState|S];
+        [_|_] ->
+            erlang:error(#{issue => config_state_failed,
+                           cluster_nodes => Nodes,
+                           failed_nodes => F})
     end.
 
-mk_node_start_filter() ->
-    fun(#local_config{key = node_start}) ->
-        false;
-       (_) ->
-        true
-    end.
+compute_config_file_version() ->
+    ConfigFile = get_ejabberd_config_path(),
+    State = parse_file(ConfigFile),
+    mongoose_config:compute_config_file_version(State).
+
+config_info() ->
+    [{config_file_version, compute_config_file_version()},
+     {config_version, compute_config_file_version()}].
```

Comment:
Shouldn't be the version of config in memory computed here?

Reply:
true.